federatedscope/core/fed_runner.py

```diff
@@ -461,7 +461,8 @@ def _run_simulation(self):
         server_msg_cache = list()
         while True:
             if len(self.shared_comm_queue) > 0:
-                msg = self.shared_comm_queue.popleft()
+                # msg = self.shared_comm_queue.popleft()
```

Comment:
Why do we need to do the `pop` operation after handling the message here? Can you give an example?

Reply:
This modification was applied to increase robustness of the pipeline, in both standalone and distributed settings, it is generally a better practice (in my opinion) to remove the query only after it has been fully handled. 
One example would be for a distributed algorithm where one client process has been killed during training, in this case, the run can continue when the client restarts and receive the message again. 
Another example is when the standalone process is paused (such as in Jupyter), the experiment can continue by rerunning function `runner.run()`, the algorithm will continue from the last message still left in queue.
On the other hand, if the message is first popped then handled, the runner is unable to recover, since it reads an empty queue and considers the run finished.

Reply:
- `_run_simulation` is only used in standalone mode.
- Please double check the case in Jupyter @rayrayraykk 
- The main idea behind the original implementation (i.e., pop first and handle later) is to ensure the popped msg is exactly the one you handled, since during the handling process some new message might be generated and pushed into the comm_queue